Nhse o34 orkv.i84 querygrant

martinsumner opened the pull request:
Enforce authorisation checks on all HTTP APIs, but allow the previous behaviour to be supported by non-default configuration.

Document the availability of security controls in Riak Quick Docs.

Tested in https://github.com/OpenRiak/riak_test/pull/16
Resolves https://github.com/OpenRiak/riak_kv/issues/84

martinsumner committed:
Add configurable support for exempting ops requests for authorisation

Default will change, to enforce authorisation on all requests when security i s enabled

martinsumner committed:
Add security section to Ops guide

martinsumner committed:
Slight doc change

WarpEngineer commented on docs/OperationsAndTroubleshootingGuide.md:
```diff
@@ -262,7 +262,125 @@ Each worker pool will regularly log its current queue length and last checkout t
-> TODO - Section on applying authentication and authorisation 
+> Riak is expected to be deployed into secure environments, it is not a database designed for direct exposure on public networks.
+
+Riak does have the optional capability to enable additional security controls, controls which are disabled by default.  However:
```
```suggestion
Riak does have the optional capability to enable additional security controls, which are disabled by default.  However:
```

WarpEngineer commented on docs/OperationsAndTroubleshootingGuide.md:
```diff
@@ -262,7 +262,125 @@ Each worker pool will regularly log its current queue length and last checkout t
+
+### Enabling Security and Restricting Source
+
+To provide further security, the Riak security mechanism should be enabled. This is not possible through configuration, it must be enabled via the command line:
```
```suggestion
To provide further security, the Riak security mechanism should be enabled. This is not possible through configuration; it must be enabled via the command line:
```

WarpEngineer commented on docs/OperationsAndTroubleshootingGuide.md:
```diff
@@ -262,7 +262,125 @@ Each worker pool will regularly log its current queue length and last checkout t
+
+This would permit access to the APIs only from the IP address `192.168.6.7` (this may be the address of a web application firewall, for example), and trust all access from that source as long as the username of `proxy_waf` is provided within the Authorization header.
+
+> In this simple case, this is functionally equivalent to applying an IP filter on the node through a standard filter utility, but it is not the security equal of that measure.  An IP filter would prevent connections being made from an unauthorised host, whereas the Risk security control allows connections and requests to be sent, but blocks requests during the processing of those requests; which ...
```
```suggestion
> In this simple case, this is functionally equivalent to applying an IP filter on the node through a standard filter utility, but it is not the security equal of that measure.  An IP filter would prevent connections being made from an unauthorised host, whereas the Riak security control allows connections and requests to be sent, but blocks requests during the processing of those requests; which presents a broader attack surface.
```

WarpEngineer commented on docs/OperationsAndTroubleshootingGuide.md:
```diff
@@ -262,7 +262,125 @@ Each worker pool will regularly log its current queue length and last checkout t
+- `certificate`;
+  - Supported for **the PB API only**,
+  - Requires that the username match the certificate name,
+  - By inference requires the session to have included a valid client certificate in the TLS negotiation.
```
May need to mention that the client cert must be signed by the same CA as the server cert.

Also, should we mention PAM and mark it as deprecated here?

martinsumner replied:
Updated - https://github.com/OpenRiak/riak_kv/pull/85/commits/e091db55d7286ad0cfea32730942e2e9dfd2b5d4

WarpEngineer commented on docs/OperationsAndTroubleshootingGuide.md:
```diff
@@ -262,7 +262,125 @@ Each worker pool will regularly log its current queue length and last checkout t
+
+For all other API endpoints, only `source` protection is applied.
+
+> With the PB API, authentication is provided at the start of a connection, and grants are assessed and cached for that connection to be used against each request.  One the HTTP API, each request on a connection is authenticated and has grant checks made independently on a per-request basis.
```
```suggestion
> With the PB API, authentication is provided at the start of a connection, and grants are assessed and cached for that connection to be used against each request.  On the HTTP API, each request on a connection is authenticated and has grant checks made independently on a per-request basis.
```

martinsumner committed:
Apply suggestions from code review

Co-authored-by: A. G. Madi <WarpEngineer@users.noreply.github.com>

martinsumner committed:
Update to security source doc following review

Added a caveat to indicate that user/password are not really well named.  There's basic protection of passwords through hashing etc, but probably not the protection that might be expected for managing the login credentials of real people.
